### ISSUE TITLE
Multiple Autonomy Nodes

### ISSUE THREAD

**Issue by StefanCaldararu** [uwsbel/autonomy-research-testbed#80]
The Issue: Different vehicles will likely want to have different implementations for the autonomy nodes (Perception, Control, SE, PP). How do we go about this?

**Comment by StefanCaldararu** [comment]
@AaronYoung5's idea: We have multiple nodes in each folder, and a different one gets used depending on the metapackage that is used (ex. multiple control nodes, but only one gets built & used). 

Why I'm not the biggest fan: Some of these will look very different, and some will be almost identical. For example, there could just be a different between the control algorithm used. For this it doesn't make sense to have a different file, it should just be passed as a parameter when you run ros. on the other hand, some controllers take different inputs, and as a result the control node must subscribe to different topics (ex. if you want to do back propagation). From what I see, this means we either have two different levels of abstraction for one issue, or a lot of duplicate files. 

@AaronYoung5 thoughts?

**Comment by AaronYoung5** [comment]
Ya I see each node having a different level of abstraction. So if you can incorporate another control algorithm in an existing one (like making the current P controller a PI controller), then do it. If a node should take additional inputs (like maybe the PI controller needs perception for some reason), then add it conditionally. if it gets to the point where the logic is sufficiently different (like MPC), then make a new node/package. And lastly, if there is common code that is shared across sufficiently different nodes, we make it an importable package that other files can reference.

So TL;DR I think we should try our best to have a main that has _all_ nodes. We then select the nodes we want through metapackages/launch files.

I agree this may get confusing, and is more on a case-by-case basis. But what would be the alternative?

**Comment by StefanCaldararu** [comment]
> I agree this may get confusing, and is more on a case-by-case basis. But what would be the alternative?

No clue, I'm just complaining about this solution lol. I'll think about it more, and I think it will become more clear as Harry and I merge the MPC / NN controllers into master...

**Comment by StefanCaldararu** [comment]
@AaronYoung5 what should htese new packages be called? for example, we ahve a path planner that outputs the error state, and takes the current state as input (very different from the path planner for the cone detectors, so makign a new node / package). Should this be called `art_oak_path_planner`?

**Comment by AaronYoung5** [comment]
@StefanCaldararu It should be relevant to the specific algorithm. See [autoware.universe](https://github.com/autowarefoundation/autoware.universe) for inspo. And not this is consistent with [ROS best practices](https://ros.org/reps/rep-0144.html).

**Comment by StefanCaldararu** [comment]
@AaronYoung5 How about the main launch file? should everyhting launch from `art_launch` and then arguments are passed, like `vehicle:=oak` or `mission:=waypoints` or something like that? or do we create a full enw launch package which is `art_launch_waypoints` or `art_launch_oak` or something?

**Comment by AaronYoung5** [comment]
Hmm getting harry (pun intended). Uhh maybe a top level `art_oak_launch` in the `art_launch` package? This is probs diverging from the ROS standard cause I don't think it's usually common to have multiple vehicle types live in the same repo. 

Ya I think `art.launch.py` should be renamed to `art_dev.launch.py` and then you create one called `art_oak.launch.py` and other vehicles follow like that. How does that sound?

**Comment by AaronYoung5** [comment]
But like logic (like common launch configurations, containers, etc.) should be added to a common launch file or shared package, idk.

**Comment by StefanCaldararu** [comment]
> But like logic (like common launch configurations, containers, etc.) should be added to a common launch file or shared package, idk.

wdym? like how so, more than ahving launchfiles for nodes / groups of nodes?

**Comment by StefanCaldararu** [comment]
@AaronYoung5 how about for sensing? No one is going to build ALL of the sensing packages, why do we even need this? why not jsut ahve each call `art_[VEH]_meta` include the sensors it needs?

**Comment by AaronYoung5** [comment]
Can you explain a little bit more what you're thinking? Like outline the commands and/or file structure you're thinking

**Comment by StefanCaldararu** [comment]
atm there is the `art_sensing` package. this currently includes `usb_cam

**Comment by StefanCaldararu** [comment]
atm there is the `art_sensing` package. This has `usb_cam` and `bluespace_mti_driver` in it. The way it is structured now, it seems that every sensor we add should go in the `art_sensing` package, but this doesn't make sense to me. `oak` is going to have a very specific sensor suite, that will likely not overlap almost at all with any of the other packages. What I see making sense is `art_sensing` is the parallel of `art_dev`, meaning that it never actually gets built, but includes **all** of the sensing packages, so that people can reference it. The other solution I see as being feasible is that each vehicle has the equivalent of `art_oak_sensing`, but this is just an extra layer that doesn't really seem necessary to me. Thoughts? Any structure that might be better than these?

**Comment by AaronYoung5** [comment]
you love the close with comment button huh

`art_sensing` isn't actually a package, like `art_sensing` is never going to be in an `exec_depend` (it doesn't have a `package.xml`). Instead, the packages with which provide sensing are stored there. I don't get what you mean by `art_dev`, I don't think there's a package called that. I think all sensing packages should go in the `art_sensing` subfolder and we use `exec_depend` to conditionally build/link to the sensing packages we actually need for a given vehicle.

**Comment by StefanCaldararu** [comment]
Solved by metapackages.
